state/application.go

```diff
@@ -1410,6 +1423,60 @@ func (a *Application) SetCharm(cfg SetCharmConfig) (err error) {
 	return nil
 }
 
+// preUpgradeRelationLimitCheck ensures that the already established relation
+// counts do not violate the max relation limits specified by the charm version
+// we are attempting to upgrade to.
+func (a *Application) preUpgradeRelationLimitCheck(newCharm *Charm) error {
+	var (
+		existingRels []*Relation
+		err          error
+	)
+
+	for relName, relSpec := range newCharm.Meta().CombinedRelations() {
+		if relSpec.Limit == 0 {
```

Comment:
`relSpec` has `optional`[1], do we count if this is set to true?

1. https://github.com/juju/charm/blob/v6/meta.go#L884

Reply:
This is  merely a hint that the charm can work even when the relation is not established. The above code only cares for already established relations. For more info see [here](https://discourse.juju.is/t/implementing-relations/1051):

```
optional is ignored by Juju, but if present should only be set to true, on requires relations, to indicate that the charm can still function effectively without having those relations added.
```